bridgenode/server.go

```diff
@@ -5,15 +5,83 @@ import (
 	"fmt"
 	"math"
 	"net"
+	"os"
+	"time"
 
+	"github.com/btcsuite/btcd/chaincfg"
 	"github.com/mit-dci/utreexo/util"
 	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
 )
 
+func ArchiveServer(param chaincfg.Params, dataDir string, sig chan bool) error {
+
+	// Channel to alert the tell the main loop it's ok to exit
+	haltRequest := make(chan bool, 1)
+
+	// Channel for ServeBlock() to wait
+	haltAccept := make(chan bool, 1)
+
+	// Handle user interruptions
+	go stopServer(sig, haltRequest, haltAccept)
+
+	_, err := os.Stat(dataDir)
+	if os.IsNotExist(err) {
+		return fmt.Errorf("%s not found, can't serve blocks\n", dataDir)
+	}
+
+	// TODO ****** server shouldn't need levelDB access, fix this
+	ttlpath := "utree/" + param.Name + "ttldb"
+	// Open leveldb
+	o := opt.Options{CompactionTableSizeMultiplier: 8}
+	lvdb, err := leveldb.OpenFile(ttlpath, &o)
+	if err != nil {
+		fmt.Printf("initialization error.  If your .blk and .dat files are ")
+		fmt.Printf("not in %s, specify alternate path with -datadir\n.", dataDir)
+		return err
+	}
+	defer lvdb.Close()
+	// **********************************
+
+	// Init forest and variables. Resumes if the data directory exists
+	maxHeight, err := restoreHeight()
+	if err != nil {
+		return err
+	}
+
+	blockServer(maxHeight, dataDir, haltRequest, haltAccept, lvdb)
+
+	return nil
+}
+
+// stopServer listens for the signal from the OS and initiates an exit sequence
+func stopServer(sig, haltRequest, haltAccept chan bool) {
+
+	// Listen for SIGINT, SIGQUIT, SIGTERM
+	<-sig
+	haltRequest <- true
+	// Sometimes there are bugs that make the program run forever.
+	// Utreexo binary should never take more than 10 seconds to exit
+	go func() {
+		time.Sleep(2 * time.Second)
+		fmt.Println("Exit timed out. Force quitting.")
+		os.Exit(1)
+	}()
+
+	// Tell the user that the sig is received
+	fmt.Println("User exit signal received. Exiting...")
+
+	// Wait until server says it's ok to exit
+	<-haltAccept
+	os.Exit(0)
+}
+
 // blockServer listens on a TCP port for incoming connections, then gives
 // ublocks blocks over that connection
 func blockServer(endHeight int32, dataDir string, haltRequest,
 	haltAccept chan bool, lvdb *leveldb.DB) {
+	endHeight--
+	fmt.Printf("serving up to block height %d\n", endHeight)
```

Comment:
maybe it would be helpful to say "up to and including", just to be explicit

Reply:
OK changed it... and the line before